torch/_tensor_docs.py

```diff
@@ -3201,11 +3210,15 @@ def callable(a, b) -> number
 :attr:`self`, :attr:`index` and :attr:`src` should have same number of
 dimensions. It is also required that ``index.size(d) <= src.size(d)`` for all
 dimensions ``d``, and that ``index.size(d) <= self.size(d)`` for all dimensions
-``d != dim``.
+``d != dim``. Note that ``index`` and ``src`` do not broadcast.
 
 Note:
     {forward_reproducibility_note}
 
+.. warning::
+
+    When indices are not unique, the gradient will be incorrect!
```

Comment:
@nikitaved I thought for scatter_add even with non-unique indices the gradient will be correct? The gradient will be propagated to all source locations, as it should be.

Reply:
scatter_add is correct only for the cases when `src.shape <= target.shape`, because of the `gather` restriction.
When `src.shape > target.shape`, `gather will throw an error about inconsistent shapes. What should be done, is that the values that are not touched by `gather` in backward has to be set to zero, that for all indices of `src` which are outside of `index.shape`.

Reply:
Thanks. If it's correct in all cases that don't throw an error then this warning should be deleted. I had based it on https://github.com/pytorch/pytorch/issues/31776#issuecomment-570581487, which said uniqueness is required.

Reply:
Both scatter and scatter_add have issues with handling gradient when `src.shape > index.shape`, but `scatter` has an additional issue of being non-deterministic with non-unique indices.